# Inpactor2: A software based on deep learning to identify and classify LTR-retrotransposons in plant genomes
It is a software based on four neural networks, whose input is an assembled genome and output is a library of LTR-retrotransposons. Optionally, Inpactor2 can run an annotation of the library with RepeatMasker. This tool was developed with the aim of decreasing run times of LTR-retrotransposon analyses on plant genomes and thus speeding up large-scale genomic analyses (e.g. those using many plant species).

First, Inpactor2 receives a genome assembly in FASTA format, then each sequence in the input file is split into windows of 50 Kb, without overlapping (stride of 50 Kb). Each window is converted into a 2D-representation using one-hot encoding. Next, a  convolutional neural network (CNN) called "Inpactor2_Detect" is used to predict which windows contain LTR-RTs and these windows are retained for further analysis. Then, LTR_finder is run to search the beginning and end positions of the previously detected LTR-RT. This step is executed in parallel in order to reduce the execution time. After, a CNN called "Inpactor2_K-mers" is used to count k-mer frequencies in the extracted LTR-retrotransposons. This CNN is intended to extract features required by the next NNs of the pipeline in a time-efficient way. Then, intact sequences are filtered and retained based on the k-mer frequencies and a fully connected neural network (FNN) called "Inpactor2_Filter". Finally, another FNN named "Inpactor2_Class" is used to classify the elements into lineages. 

Inpactor2 can be executed using several (from one to five) cycles of analysis to detect LTR-RTs divided by the windows generated by the software. Additionally, Inpactor2 can be performed using different structural parameters to filter LTR-RTs, such as minimum and maximum length, LTR beginning with TG and finishing with CA, and target site duplication (TSD) before and after the element.

![alt text](https://github.com/simonorozcoarias/Inpactor2/blob/main/NN_architectures/Inpactor_structure.png)

## Installation:
We highly recommend to use and install Python packages within an Anaconda environment. First, download the lastest version of Inpactor2
```
git clone https://github.com/simonorozcoarias/Inpactor2.git
```
Go to the Inpactor2 folder and find the file named "Inpactor2.yml". Then, install de environment: 
```
conda env create -f Inpactor2.yml
```

## Usage:
Activate the anaconda environment:
```
conda activate Inpactor2
```
Then execute:
```
python3 Inpactor2.py -f <input_genome_file.fasta> [
```
